argo/src/Argo/ServerState.hs

```diff
@@ -0,0 +1,180 @@
+stateRecipes = lens _stateRecipes (\s r -> s { _stateRecipes = r })
+
+appStateCache :: Lens' (ServerState recipeStep s) (IORef (HashMap [recipeStep] s))
+appStateCache = lens _appStateCache (\s c -> s { _appStateCache = c })
+
+initialAppState :: Lens' (ServerState recipeStep s) s
+initialAppState = lens _initialAppState (\s a -> s { _initialAppState = a })
+
+stateFiles :: Lens' (ServerState recipeStep s) (IORef (HashMap ([recipeStep], FilePath) FileHash))
+stateFiles = lens _stateFiles (\s fs -> s { _stateFiles = fs })
+
+stateFileContents :: Lens' (ServerState recipeStep s) (IORef (HashMap FileHash ByteString))
+stateFileContents = lens _stateFileContents (\s fc -> s { _stateFileContents = fc })
+
+stateRecipe :: ServerState recipeStep s -> StateID -> IO (Maybe [recipeStep])
+stateRecipe server sid =
+  do recipes <- readIORef $ view stateRecipes server
+     pure $ HM.lookup sid recipes
+
+
+recipeFile ::
+  (Eq recipeStep, Hashable recipeStep) =>
+  ServerState recipeStep s ->
+  [recipeStep] ->
+  FilePath -> IO ByteString
+recipeFile server recipe filename =
+  do hashes <- readIORef $ view stateFiles server
+     theHash <-
+       case view (at (recipe, filename)) hashes of
+         Nothing ->
+           ioError $
+           IOError { ioe_handle = Nothing
+                   , ioe_type = NoSuchThing
+                   , ioe_location = "restoring " ++ filename
+                   , ioe_description = "No such cached file " ++ filename
+                   , ioe_errno = Nothing
+                   , ioe_filename = Just filename
+                   }
+         Just h -> pure h
+     fileContents <- readIORef $ view stateFileContents server
+     case view (at theHash) fileContents of
+       Nothing ->
+         ioError $
+         IOError { ioe_handle = Nothing
+                 , ioe_type = NoSuchThing
+                 , ioe_location = "restoring " ++ filename
+                 , ioe_description = "No contents for cached file " ++ filename
+                 , ioe_errno = Nothing
+                 , ioe_filename = Just filename
+                 }
+
+       Just bs -> pure bs
+
+
+getAppState ::
+  (Eq recipeStep, Hashable recipeStep) =>
+  (s -> recipeStep -> (FilePath -> IO ByteString) -> IO s) ->
+  ServerState recipeStep s ->
+  [recipeStep] ->
+  IO s
+getAppState runStep server recipe =
+  do recipeCache <- readIORef $ view appStateCache server
+     case view (at recipe) recipeCache of
+       Just s -> return s
+       Nothing -> runRecipe runStep server recipe
+
+runRecipe ::
+  (Eq recipeStep, Hashable recipeStep) =>
+  (s -> recipeStep -> (FilePath -> IO ByteString) -> IO s) ->
+  ServerState recipeStep s ->
+  [recipeStep] ->
+  IO s
+runRecipe _ server [] = pure $ view initialAppState server
+runRecipe runStep server recipe@(step : steps) =
+  do s <- getAppState runStep server steps
+     let fileCache = recipeFile server recipe
+     runStep s step fileCache
```

Comment:
It took me a second to parse out the mutual recursion between `getAppState` and `runRecipe`. Is there a way to phrase this that isn't mutually recursive? If not, it might be good to comment describing how they interact (I always find mutual recursion takes me a hot sec to load up in my head!).

Reply:
We could do it by inlining them. Basically, one of them checks whether the recipe's state is cached, and the other reconstructs it if not. 

I'll at least add some docs.